hathor/cli/db_export.py

```diff
@@ -0,0 +1,99 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+import struct
+from argparse import ArgumentParser, FileType, Namespace
+from typing import Iterator, Optional
+
+from hathor.cli.run_node import RunNode
+
+MAGIC_HEADER = b'HathDB'
+
+
+class DbExport(RunNode):
+    def start_manager(self, args: Namespace) -> None:
+        pass
+
+    def register_signal_handlers(self, args: Namespace) -> None:
+        pass
+
+    def register_resources(self, args: Namespace) -> None:
+        pass
+
+    def create_parser(self) -> ArgumentParser:
+        from hathor.conf import HathorSettings
+        settings = HathorSettings()
+
+        def max_height(arg: str) -> Optional[int]:
+            if arg.lower() == 'checkpoint':
+                if not settings.CHECKPOINTS:
+                    raise ValueError('There are no checkpoints to use')
+                return settings.CHECKPOINTS[-1].height
+            elif arg:
+                return int(arg)
+            else:
+                return None
+
+        parser = super().create_parser()
+        parser.add_argument('--export-file', type=FileType('wb', 0), required=True,
+                            help='Save the export to this file')
+        parser.add_argument('--export-iterator', choices=['fast2', 'fast', 'slow'], default='fast2',
+                            help='Which method of iterating to use, don\'t change unless you know what it does')
+        parser.add_argument('--export-max-height', type=max_height,
+                            help='Stop exporting after a certain height, use "checkpoint" stop at the last one')
+        return parser
+
+    def prepare(self, args: Namespace) -> None:
+        from hathor.transaction import BaseTransaction
+
+        super().prepare(args)
+
+        # allocating io.BufferedWriter here so we "own" it
+        self.out_file = io.BufferedWriter(args.export_file)
+
+        self.iter_tx: Iterator[BaseTransaction]
+        if args.export_iterator == 'fast2':
+            self.iter_tx = self.tx_storage._topological_fast2()
+        elif args.export_iterator == 'fast':
+            self.iter_tx = self.tx_storage._topological_fast()
+        elif args.export_iterator == 'slow':
+            self.iter_tx = self.tx_storage._topological_sort()
+        else:
+            raise ValueError(f'unknown iterator "{args.export_iterator}"')
+
+        self.export_height = args.export_max_height
+
+    def run(self) -> None:
+        from hathor.util import progress
+        self.out_file.write(MAGIC_HEADER)
```

Comment:
Should we add some extra information such as the total number of transactions and blocks?

Reply:
That makes sense, I'll do that.

Reply:
Nice implementation. Good job!